### PR TITLE
Feat: Add support for provider hooks

### PR DESCRIPTION
**Pull request by benjiro**
Implement provider hook spec https://github.com/open-feature/spec/pull/119

- [Breaking] Remove IFeatureProvider interface in favor of FeatureProvider abstract class
so default implementations can exist
- Make sure Provider hooks are called in the correct order
- Use strick mocking mode so sequence is validated correctly
- Update test cases for provider hooks support

### CODE REVIEW

File: src/OpenFeature.SDK/OpenFeature.cs [hunk 1]
```diff
@@ -81,7 +81,7 @@ public sealed class OpenFeature
         /// Sets the global <see cref="EvaluationContext"/>
         /// </summary>
         /// <param name="context"></param>
-        public void SetContext(EvaluationContext context) => this._evaluationContext = context;
+        public void SetContext(EvaluationContext context) => this._evaluationContext = context ?? new EvaluationContext();
```
Comment (toddbaert) [inline src/OpenFeature.SDK/OpenFeature.cs:84]: `??` - features I didn't know about in C#7

Very nice. https://docs.microsoft.com/en-in/dotnet/csharp/language-reference/operators/null-coalescing-operator

File: src/OpenFeature.SDK/OpenFeatureClient.cs [hunk 2]
```diff
@@ -209,6 +209,7 @@ public FeatureClient(IFeatureProvider featureProvider, string name, string versi
                 .Concat(OpenFeature.Instance.GetHooks())
                 .Concat(this._hooks)
                 .Concat(options?.Hooks ?? Enumerable.Empty<Hook>())
+                .Concat(this._featureProvider.GetProviderHooks())
```
Comment (toddbaert) [inline src/OpenFeature.SDK/OpenFeatureClient.cs:212]: I think the merging works as expected here (basically, hooks win... see: https://github.com/open-feature/spec/blob/main/specification/sections/04-hooks.md#requirement-434)

But I _don't_ see a test that specifically asserts this behavior (let me know if I'm wrong). If I'm correct, and it's alright with you @benjiro , I will write this test myself just to make sure I can get this project setup fully locally (I've pulled it but never run the tests suite).

**Review by toddbaert — APPROVED** [review]
LGTM, let me know about comments on the merge behavior change.

File: src/OpenFeature.SDK/OpenFeatureClient.cs (hunk 2 above)
Comment (benjiro) [reply src/OpenFeature.SDK/OpenFeatureClient.cs:212]: Yea go ahead and update the test case. Let me know if you have any issues I've only tested the project via command line, Rider and VS. if you're using vscode you might need to fix up the git ignores or do some initial setup. If you run into any troubles ping me on slack happy to help

File: src/OpenFeature.SDK/OpenFeature.cs (hunk 1 above)
Comment (benjiro) [reply src/OpenFeature.SDK/OpenFeature.cs:84]: Now that the dotnet team do a major release every year the language is evolving at a pretty fast rate 👍🏻

**Comment by benjiro** [comment]
@toddbaert fine with you tackling the test, did you want to merge this PR or add the test to it?

File: src/OpenFeature.SDK/OpenFeatureClient.cs (hunk 2 above)
Comment (toddbaert) [reply src/OpenFeature.SDK/OpenFeatureClient.cs:212]: I'm going to branch of this branch to make this change, feel free to merge or not.

**Comment by toddbaert** [comment]
> @toddbaert fine with you tackling the test, did you want to merge this PR or add the test to it?

Go ahead and merge, I'll do it in a separate PR. Will start on that today.

**Pull request merged**